stl/inc/span

```diff
@@ -0,0 +1,678 @@
+// clang-format off
+template <class _Cont, class _Ty>
+struct _Is_span_compatible_container : conjunction<
+	negation<is_array<_Cont>>,
+	negation<_Is_span<remove_const_t<_Cont>>>,
+	negation<_Is_std_array<remove_const_t<_Cont>>>,
+	_Has_container_interface<_Cont>,
+	_Is_span_convertible_range<_Cont, _Ty>> {};
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using _Mybase                = _Span_extent_type<_Extent>;
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    template <size_t _Ext = _Extent, enable_if_t<_Ext == 0 || _Ext == dynamic_extent, int> = 0>
+    constexpr span() noexcept {};
+
+    constexpr span(pointer _Ptr, index_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_Ptr) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Count == _Extent, "span construction from invalid range as count != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    constexpr span(pointer _First, pointer _Last) noexcept // strengthened
+        : _Mybase(static_cast<index_type>(_Last - _First)), _Mydata(_First) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_First <= _Last, "span construction from pointers only valid for first <= last");
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Last - _First == _Extent, "span construction from invalid range as last - first != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(element_type (&_Arr)[_Size]) noexcept : _Mybase(_Size), _Mydata(_Arr) {}
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    // clang-format off
+    template <size_t _Size, size_t _Ext = _Extent,
+        enable_if_t<conjunction_v<is_convertible<const value_type*, element_type*>,
+                                  bool_constant<_Ext == dynamic_extent || _Ext == _Size>>, int> = 0>

```

Comment:
It probably doesn't matter for throughput, but I observe that in these conjunctions, it would almost certainly be better to put the `bool_constant` first, since that's a cheap comparison of integers. If that short-circuits, we can avoid the more expensive `is_convertible` etc. check.

Reply:
I plan on tackling that this week. With the previous `_Span_iterator` version I had some problems with comparability and decided to postpone them after the rewrite.

Shockingly I forgot and in my iterator [tests ](https://github.com/miscco/SpanTest/blob/540d867a8e478f4430e97e0c3f7c91e02ee045f5/SpanTest/begin.cpp#L52) the comparability is commented out ...

Reply:
Done